### PR TITLE
Resolve: "Expand subgraphs into parent graph"

### LINKED ISSUE #64: Expand subgraphs into parent graph
In GitLab by @mariusalexander on Feb 16, 2024, 13:48

One can create a subgraph by grouping nodes. It would be nice if one could expand the subgraph into the parent graph, effectively undoing the grouping action

### PR DESCRIPTION
**Pull request by mariusalexander**
Closes #64

### CODE REVIEW

File: src/intelli/graph.h [hunk 1]
```diff
@@ -375,6 +420,110 @@ class GT_INTELLI_EXPORT Graph : public Node
...
+     * updated. Failing to move a node will not restore the previous state
+     * of the graph. Use a MementoDiff for this purpose. The underlying node
+     * object is guranteed to be kept alive.
+     * Note: All connections to this node are destroyed.
+     * @param nodeId Node to move to other graph
+     * @param policy Whether to update the node's id if necessary (if the
+     * node id should be kept, moving the node may fail)
+     * @return success
+     */
+    bool moveNode(NodeId nodeId,
```
Comment (jensschmeink) [inline src/intelli/graph.h:435]: Why is the implementation in the header? We normaly only do this for template functions.

File: src/intelli/graph.h [hunk 2]
```diff
@@ -375,6 +420,110 @@ class GT_INTELLI_EXPORT Graph : public Node
...
+     * @return success
+     */
+    template <typename NodeList>
+    bool moveNodesAndConnections(NodeList const& nodes,
+                                 Graph& targetGraph,
+                                 NodeIdPolicy policy = NodeIdPolicy::Update)
+    {
+        auto changeThis   = modify();
+        auto changeTarget = targetGraph.modify();
+        Q_UNUSED(changeThis);
```
Comment (jensschmeink) [inline src/intelli/graph.h:474]: why do you set up variables and do not use them?

File: src/intelli/graph.h [hunk 3]
```diff
@@ -375,6 +420,110 @@ class GT_INTELLI_EXPORT Graph : public Node
...
+     * @return success
+     */
+    template <typename NodeList>
+    bool moveNodes(NodeList const& nodes,
+                   Graph& targetGraph,
+                   NodeIdPolicy policy = NodeIdPolicy::Update)
+    {
+        auto changeThis   = modify();
+        auto changeTarget = targetGraph.modify();
+        Q_UNUSED(changeThis);
```
Comment (jensschmeink) [inline src/intelli/graph.h:518]: same here?

File: src/intelli/graphconnectionmodel.h [hunk 4]
```diff
@@ -1019,6 +1020,12 @@ class ConnectionModel_t
 
     bool containts(key_type const& key) const { return m_data.contains(key); }
 
+    bool empty() const { return m_data.empty(); }
+    bool isEmpty() const { return m_data.empty(); }
+
+    size_type size() const { return m_data.size(); }
```
Comment (jensschmeink) [inline src/intelli/graphconnectionmodel.h:1026]: Why size_type and not size_t?

File: src/intelli/gui/graphscene.cpp [hunk 5]
```diff
@@ -1423,20 +1321,143 @@ GraphScene::groupNodes(QVector<NodeGraphicsObject*> const& selectedNodeObjects)
...
+
+    auto* inputProvider  = groupNode->inputProvider();
+    auto* outputProvider = groupNode->outputProvider();
+    assert(inputProvider);
+    assert(outputProvider);
+
+    // gather input and output connections
+    QVector<ConnectionUuid> expandedInputConnections, expandedOutputConnections;
+
+    {
```
Comment (jensschmeink) [inline src/intelli/gui/graphscene.cpp:1362]: please comment this block to make clear why to use new scope

File: src/intelli/gui/graphscene.cpp [hunk 6]
```diff
@@ -1423,20 +1321,143 @@ GraphScene::groupNodes(QVector<NodeGraphicsObject*> const& selectedNodeObjects)
...
+                assert(sourceNode);
+                gtDebug() << sourceNode->caption() << "TO" << groupNode->findNode(conId.inNodeId)->caption();
+                conUuid.outNodeId = sourceNode->uuid();
+                conUuid.outPort   = connection.port;
+
+                expandedInputConnections.push_back(conUuid);
+            }
+        }
+
+        gtDebug() << "SWITCH";
```
Comment (jensschmeink) [inline src/intelli/gui/graphscene.cpp:1385]: Please give a more meaningful message here.

**Review by jensschmeink — COMMENTED** [review]
Added some questions about the code

File: src/intelli/graph.h (hunk 2 above)
Comment (mariusalexander) [reply src/intelli/graph.h:474]: This is a scoped object, that basically stops execution updates while it is alive. I thus only have to instantiate it and do not have to use it again (It is a `gt::finally` function)

Comment (jensschmeink) [reply src/intelli/graph.h:474]: Ok. Then please do not use auto here to make it clear to the reader

File: src/intelli/gui/graphscene.cpp (hunk 6 above)
Comment (mariusalexander) [reply src/intelli/gui/graphscene.cpp:1385]: thanks, this must be removed

File: src/intelli/graph.h (hunk 2 above)
Comment (mariusalexander) [reply src/intelli/graph.h:474]: I'll rename the variable and replace auto with its type

File: src/intelli/graphconnectionmodel.h (hunk 4 above)
Comment (mariusalexander) [reply src/intelli/graphconnectionmodel.h:1026]: In this class I use QHash. Any container that should be compatible with the STL should define a `size_type` variable. Here I use forwarded `QHash::size_type` and I thought why not use it here as well.

File: src/intelli/graph.h (hunk 1 above)
Comment (mariusalexander) [reply src/intelli/graph.h:435]: I guess I can move it into the .cpp as well 👍

**Pull request merged**